RecoTracker/LSTCore/interface/alpaka/Module.h

```diff
@@ -0,0 +1,372 @@
+
+    template <typename TQueue, typename TDevAcc>
+    objectRangesBuffer(unsigned int nMod, unsigned int nLowerMod, TDevAcc const& devAccIn, TQueue& queue)
+        : hitRanges_buf(allocBufWrapper<int>(devAccIn, nMod * 2, queue)),
+          hitRangesLower_buf(allocBufWrapper<int>(devAccIn, nMod, queue)),
+          hitRangesUpper_buf(allocBufWrapper<int>(devAccIn, nMod, queue)),
+          hitRangesnLower_buf(allocBufWrapper<int8_t>(devAccIn, nMod, queue)),
+          hitRangesnUpper_buf(allocBufWrapper<int8_t>(devAccIn, nMod, queue)),
+          mdRanges_buf(allocBufWrapper<int>(devAccIn, nMod * 2, queue)),
+          segmentRanges_buf(allocBufWrapper<int>(devAccIn, nMod * 2, queue)),
+          trackletRanges_buf(allocBufWrapper<int>(devAccIn, nMod * 2, queue)),
+          tripletRanges_buf(allocBufWrapper<int>(devAccIn, nMod * 2, queue)),
+          trackCandidateRanges_buf(allocBufWrapper<int>(devAccIn, nMod * 2, queue)),
+          quintupletRanges_buf(allocBufWrapper<int>(devAccIn, nMod * 2, queue)),
+          nEligibleT5Modules_buf(allocBufWrapper<uint16_t>(devAccIn, 1, queue)),
+          indicesOfEligibleT5Modules_buf(allocBufWrapper<uint16_t>(devAccIn, nLowerMod, queue)),
+          quintupletModuleIndices_buf(allocBufWrapper<int>(devAccIn, nLowerMod, queue)),
+          quintupletModuleOccupancy_buf(allocBufWrapper<int>(devAccIn, nLowerMod, queue)),
+          miniDoubletModuleIndices_buf(allocBufWrapper<int>(devAccIn, nLowerMod + 1, queue)),
+          miniDoubletModuleOccupancy_buf(allocBufWrapper<int>(devAccIn, nLowerMod + 1, queue)),
+          segmentModuleIndices_buf(allocBufWrapper<int>(devAccIn, nLowerMod + 1, queue)),
+          segmentModuleOccupancy_buf(allocBufWrapper<int>(devAccIn, nLowerMod + 1, queue)),
+          tripletModuleIndices_buf(allocBufWrapper<int>(devAccIn, nLowerMod, queue)),
+          tripletModuleOccupancy_buf(allocBufWrapper<int>(devAccIn, nLowerMod, queue)),
+          device_nTotalMDs_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          device_nTotalSegs_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          device_nTotalTrips_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          device_nTotalQuints_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)) {
+      alpaka::memset(queue, hitRanges_buf, 0xff);
+      alpaka::memset(queue, hitRangesLower_buf, 0xff);
+      alpaka::memset(queue, hitRangesUpper_buf, 0xff);
+      alpaka::memset(queue, hitRangesnLower_buf, 0xff);
+      alpaka::memset(queue, hitRangesnUpper_buf, 0xff);
+      alpaka::memset(queue, mdRanges_buf, 0xff);
+      alpaka::memset(queue, segmentRanges_buf, 0xff);
+      alpaka::memset(queue, trackletRanges_buf, 0xff);
+      alpaka::memset(queue, tripletRanges_buf, 0xff);
+      alpaka::memset(queue, trackCandidateRanges_buf, 0xff);
+      alpaka::memset(queue, quintupletRanges_buf, 0xff);
+      alpaka::memset(queue, quintupletModuleIndices_buf, 0xff);
+      alpaka::wait(queue);
```

Comment:
Why `wait()`?

Reply:
I'm not sure memset waits by itself.
Is the idea of the comment to keep syncronization to the higher level code?

Reply:
`alpaka::memset()` should not wait by itself. I'm wondering why this explicit synchronization between host and device is needed (because all subsequent operations enqueued in `queue` will see the result of the `memset()`). Are the buffers accessed independently of `queue`?

Reply:
ah, we have implemented a single consistent queue only recently.
I suspect the earlier implementation had to keep it for overall consistency.

Is it safe to wait here in terms of almost zero cost, or does it have to be removed?

Reply:
I think keeping the `wait()` is ok for the R&D period, but eventually it should be removed. At minimum I'd suggest to add a FIXME/TODO comment of what is going on.

Reply:
> Is it safe to wait here in terms of almost zero cost, or does it have to be removed?

It's not zero cost, and similar patterns did show in profiling runs.

Reply:
> I think keeping the `wait()` is acceptable for the R&D period, but eventually it should be removed. At minimum I'd suggest to add a FIXME/TODO comment of what is going on.

If I understand correctly, the problem is the "FIXME" is needed at the caller more importantly, if that caller currently assumes that the operations are synchronous.
E.g. in this case of a constructor without a wait, in the following use case
```C++
    rangesBuffers = new SDL::objectRangesBuffer<Dev>(nModules_, nLowerModules_, devAcc, queue);
    rangesInGPU->setData(*rangesBuffers);
```
I would need to decide to insert a `wait` or not; and without knowing the actual implementation I would have to.
Or, alternatively, the queue needs to be added to `setData` method (probably with a wait at the top).

There are a few cases in this PR where `wait` is added in the middle of the scope between calls using a `queue`. It seems straightforward to clean up those.

For the rest, is it more practical to somehow explicitly tag methods that take `queue` as an argument that they are synchronous or not?

Would you agree?

Reply:
> If I understand correctly, the problem is the "FIXME" is needed at the caller more importantly, if that caller currently assumes that the operations are synchronous. E.g. in this case of a constructor without a wait, in the following use case
> 
> ```c++
>     rangesBuffers = new SDL::objectRangesBuffer<Dev>(nModules_, nLowerModules_, devAcc, queue);
>     rangesInGPU->setData(*rangesBuffers);
> ```
> 
> I would need to decide to insert a `wait` or not; and without knowing the actual implementation I would have to. Or, alternatively, the queue needs to be added to `setData` method (probably with a wait at the top).

What makes `setData()` to expect a synchronization? As far as I can tell, it only operates on pointers (calling `alpaka::getPtrNative()` on the buffers) and doesn't de-reference the pointers (on non-CPU backends a de-reference should result a segfault). Therefore I'd think `setData()` should be fine without a `wait()` in between.

I'd expect the only case where a `wait()` is needed would be after a `alpaka::memcpy()` from device to host (if there are such intermediate memcopies called from `LSTProducer::acquire())`.

Anyway, my point with the request of "FIXME comments" was that the reason for every `wait()` that remains in this PR gets documented, so that a path to remove them can be figured out (especially if addressing those would need framework or infrastructure improvements).

> There are a few cases in this PR where `wait` is added in the middle of the scope between calls using a `queue`. It seems straightforward to clean up those.

:+1:

> For the rest, is it more practical to somehow explicitly tag methods that take `queue` as an argument that they are synchronous or not?

The framework has a convention of `Async` postfix for asynchronous functions (which is visible also in `CopyTo{Device,Host}` class template specializations). In the Alpaka-using code there is somewhat varying convention, e.g. parts of the pixel code follows the `Async` postfix, but many places more or less assume that a function with a Queue parameter would be asynchronous (there is somewhat related discussion in https://github.com/cms-sw/cmssw/pull/45572#issuecomment-2256177384).

So using naming convention could make sense (I'd be in favor of marking asynchronous functions).

Reply:
> If I understand correctly, the problem is the "FIXME" is needed at the caller more importantly, if that caller currently assumes that the operations are synchronous.

IMO if a function takes a `queue` argument, it should be assumed that the operation is asynchronous, or safe to use in an asynchronous context - that is:
  - all operations are enqueued in the given `queue`;
  - the function itself does not wait.

IMO adding `_async` or `Async` to every function that takes a `queue` argument is redundant. In fact, it's actually misleading, because if the queue is synchronous (blocking), like the CPU queue we are currently using, the operation will _not_ be asynchronous.